File: java/data_structures/linkedlist/doubly_linked_list/DoublyLinkedList.java
package data_structures.linkedlist.doubly_linked_list;

import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Objects;

/**
 * A simple, production-ready implementation of a doubly linked list.
 *
 * <p>This class is intentionally close in spirit to a minimal {@code Deque<T>}:
 * it supports adding/removing at both ends in O(1), iteration in forward
 * order, and a descending iterator.</p>
 *
 * <p>Not thread-safe.</p>
 */
public class DoublyLinkedList<T> implements Iterable<T> {

    /**
     * Internal node representation. Not exposed outside this class.
     */
    private static final class Node<E> {
        E item;
        Node<E> prev;
        Node<E> next;

        Node(Node<E> prev, E item, Node<E> next) {
            this.prev = prev;
            this.item = item;
            this.next = next;
        }
    }

    private Node<T> head;
    private Node<T> tail;
    private int size;

    /**
     * Creates an empty list.
     */
    public DoublyLinkedList() {
        // head, tail = null; size = 0;
    }

    /**
     * Creates a list populated with all elements from the given iterable,
     * in iteration order.
     */
    public DoublyLinkedList(Iterable<? extends T> elements) {
        this();
        Objects.requireNonNull(elements, "elements must not be null");
        for (T e : elements) {
            addLast(e);
        }
    }

    // ------------------------------------------------------------------
    // Basic properties / helpers
    // ------------------------------------------------------------------

    /**
     * Returns the number of elements in this list.
     */
    public int size() {
        return size;
    }

    /**
     * Returns {@code true} if this list contains no elements.
     */
    public boolean isEmpty() {
        return size == 0;
    }

    /**
     * Removes all elements from this list.
     */
    public void clear() {
        Node<T> current = head;
        while (current != null) {
            Node<T> next = current.next;
            current.prev = null;
            current.next = null;
            current.item = null;
            current = next;
        }
        head = tail = null;
        size = 0;
    }

    // ------------------------------------------------------------------
    // Core operations: add/remove/peek at both ends
    // ------------------------------------------------------------------

    /**
     * Inserts the specified element at the front of this list.
     */
    public void addFirst(T element) {
        Node<T> newNode = new Node<>(null, element, head);

        if (head == null) {
            // list was empty
            head = tail = newNode;
        } else {
            head.prev = newNode;
            head = newNode;
        }

        size++;
    }

    /**
     * Inserts the specified element at the end of this list.
     */
    public void addLast(T element) {
        Node<T> newNode = new Node<>(tail, element, null);

        if (tail == null) {
            // list was empty
            head = tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }

        size++;
    }

    /**
     * Returns, but does not remove, the first element of this list,
     * or {@code null} if the list is empty.
     */
    public T peekFirst() {
        return (head == null) ? null : head.item;
    }

    /**
     * Returns, but does not remove, the last element of this list,
     * or {@code null} if the list is empty.
     */
    public T peekLast() {
        return (tail == null) ? null : tail.item;
    }

    /**
     * Removes and returns the first element of this list.
     *
     * @throws NoSuchElementException if the list is empty
     */
    public T removeFirst() {
        if (head == null) {
            throw new NoSuchElementException("removeFirst from empty list");
        }

        Node<T> node = head;
        T value = node.item;

        if (head == tail) {
            // single element
            head = tail = null;
        } else {
            head = head.next;
            Objects.requireNonNull(head).prev = null;
        }

        // help GC
        node.next = null;
        node.item = null;
        size--;
        return value;
    }

    /**
     * Removes and returns the last element of this list.
     *
     * @throws NoSuchElementException if the list is empty
     */
    public T removeLast() {
        if (tail == null) {
            throw new NoSuchElementException("removeLast from empty list");
        }

        Node<T> node = tail;
        T value = node.item;

        if (head == tail) {
            head = tail = null;
        } else {
            tail = tail.prev;
            Objects.requireNonNull(tail).next = null;
        }

        // help GC
        node.prev = null;
        node.item = null;
        size--;
        return value;
    }

    // ------------------------------------------------------------------
    // Search / removal by value
    // ------------------------------------------------------------------

    /**
     * Returns {@code true} if this list contains at least one element
     * equal to {@code value}, using {@link Objects#equals(Object, Object)}.
     */
    public boolean contains(Object value) {
        return indexOfNode(value) != null;
    }

    /**
     * Removes the first occurrence of the specified element from this list,
     * if it is present.
     *
     * @return {@code true} if an element was removed, {@code false} otherwise
     */
    public boolean removeFirstOccurrence(Object value) {
        Node<T> node = indexOfNode(value);
        if (node == null) {
            return false;
        }
        unlink(node);
        return true;
    }

    /**
     * Internal search for a node whose item equals the given value.
     */
    private Node<T> indexOfNode(Object value) {
        for (Node<T> current = head; current != null; current = current.next) {
            if (Objects.equals(current.item, value)) {
                return current;
            }
        }
        return null;
    }

    /**
     * Unlinks the given node from the list, adjusting head/tail and size.
     * Assumes the node belongs to this list.
     */
    private void unlink(Node<T> node) {
        Node<T> prev = node.prev;
        Node<T> next = node.next;

        if (prev == null) {
            // node is head
            head = next;
        } else {
            prev.next = next;
        }

        if (next == null) {
            // node is tail
            tail = prev;
        } else {
            next.prev = prev;
        }

        node.prev = null;
        node.next = null;
        node.item = null;
        size--;
    }

    // ------------------------------------------------------------------
    // Iteration
    // ------------------------------------------------------------------

    /**
     * Returns an iterator over the elements in this list in proper sequence
     * (from first to last).
     */
    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private Node<T> current = head;
            private Node<T> lastReturned = null;

            @Override
            public boolean hasNext() {
                return current != null;
            }

            @Override
            public T next() {
                if (current == null) {
                    throw new NoSuchElementException();
                }
                lastReturned = current;
                current = current.next;
                return lastReturned.item;
            }

            @Override
            public void remove() {
                if (lastReturned == null) {
                    throw new IllegalStateException("next() has not been called or element already removed");
                }
                unlink(lastReturned);
                lastReturned = null;
            }
        };
    }

    /**
     * Returns an iterable that iterates the elements in reverse order
     * (from last to first).
     */
    public Iterable<T> descendingIterable() {
        return () -> new Iterator<T>() {
            private Node<T> current = tail;
            private Node<T> lastReturned = null;

            @Override
            public boolean hasNext() {
                return current != null;
            }

            @Override
            public T next() {
                if (current == null) {
                    throw new NoSuchElementException();
                }
                lastReturned = current;
                current = current.prev;
                return lastReturned.item;
            }

            @Override
            public void remove() {
                if (lastReturned == null) {
                    throw new IllegalStateException("next() has not been called or element already removed");
                }
                unlink(lastReturned);
                lastReturned = null;
            }
        };
    }

    // ------------------------------------------------------------------
    // toString / debug helpers
    // ------------------------------------------------------------------

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("DoublyLinkedList[");
        Node<T> current = head;
        while (current != null) {
            sb.append(String.valueOf(current.item));
            if (current.next != null) {
                sb.append(", ");
            }
            current = current.next;
        }
        sb.append(']');
        return sb.toString();
    }
}

File: csharp/data_structure/linkedlist/doubly_linked_list/DoublyLinkedList.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace DataStructures.LinkedList;

/// <summary>
/// A simple, production-ready implementation of a doubly linked list.
///
/// Supports O(1) insertion and removal at both ends, forward iteration,
/// and a reversed enumerable.
/// Not thread-safe.
/// </summary>
/// <typeparam name="T">Element type.</typeparam>
public class DoublyLinkedList<T> : IEnumerable<T>
{
    /// <summary>
    /// Internal node representation. Not exposed outside this class.
    /// </summary>
    private sealed class Node
    {
        public T Item;
        public Node? Prev;
        public Node? Next;

        public Node(Node? prev, T item, Node? next)
        {
            Prev = prev;
            Item = item;
            Next = next;
        }
    }

    private Node? _head;
    private Node? _tail;
    private int _count;

    /// <summary>
    /// Creates an empty list.
    /// </summary>
    public DoublyLinkedList()
    {
    }

    /// <summary>
    /// Creates a list populated with all elements from the given sequence,
    /// in iteration order.
    /// </summary>
    public DoublyLinkedList(IEnumerable<T> source)
    {
        if (source is null) throw new ArgumentNullException(nameof(source));
        foreach (var item in source)
        {
            AddLast(item);
        }
    }

    /// <summary>
    /// Gets the number of elements in the list.
    /// </summary>
    public int Count => _count;

    /// <summary>
    /// Returns true if the list contains no elements.
    /// </summary>
    public bool IsEmpty => _count == 0;

    /// <summary>
    /// Removes all elements from the list.
    /// </summary>
    public void Clear()
    {
        var current = _head;
        while (current is not null)
        {
            var next = current.Next;
            current.Prev = null;
            current.Next = null;
            current.Item = default!;
            current = next;
        }

        _head = _tail = null;
        _count = 0;
    }

    // --------------------------------------------------------------------
    // Core operations: add / remove / peek at both ends
    // --------------------------------------------------------------------

    /// <summary>
    /// Inserts an element at the beginning of the list.
    /// </summary>
    public void AddFirst(T value)
    {
        var newNode = new Node(prev: null, item: value, next: _head);

        if (_head is null)
        {
            // list was empty
            _head = _tail = newNode;
        }
        else
        {
            _head.Prev = newNode;
            _head = newNode;
        }

        _count++;
    }

    /// <summary>
    /// Inserts an element at the end of the list.
    /// </summary>
    public void AddLast(T value)
    {
        var newNode = new Node(prev: _tail, item: value, next: null);

        if (_tail is null)
        {
            // list was empty
            _head = _tail = newNode;
        }
        else
        {
            _tail.Next = newNode;
            _tail = newNode;
        }

        _count++;
    }

    /// <summary>
    /// Returns, but does not remove, the first element of the list.
    /// </summary>
    /// <exception cref="InvalidOperationException">If the list is empty.</exception>
    public T PeekFirst()
    {
        if (_head is null)
            throw new InvalidOperationException("PeekFirst from empty list");
        return _head.Item;
    }

    /// <summary>
    /// Returns, but does not remove, the last element of the list.
    /// </summary>
    /// <exception cref="InvalidOperationException">If the list is empty.</exception>
    public T PeekLast()
    {
        if (_tail is null)
            throw new InvalidOperationException("PeekLast from empty list");
        return _tail.Item;
    }

    /// <summary>
    /// Removes and returns the first element of the list.
    /// </summary>
    /// <exception cref="InvalidOperationException">If the list is empty.</exception>
    public T RemoveFirst()
    {
        if (_head is null)
        {
            throw new InvalidOperationException("RemoveFirst from empty list");
        }

        var node = _head;
        var value = node.Item;

        if (_head == _tail)
        {
            // single element
            _head = _tail = null;
        }
        else
        {
            _head = _head.Next;
            if (_head is not null)
            {
                _head.Prev = null;
            }
        }

        node.Next = null;
        node.Item = default!;
        _count--;
        return value;
    }

    /// <summary>
    /// Removes and returns the last element of the list.
    /// </summary>
    /// <exception cref="InvalidOperationException">If the list is empty.</exception>
    public T RemoveLast()
    {
        if (_tail is null)
        {
            throw new InvalidOperationException("RemoveLast from empty list");
        }

        var node = _tail;
        var value = node.Item;

        if (_head == _tail)
        {
            _head = _tail = null;
        }
        else
        {
            _tail = _tail.Prev;
            if (_tail is not null)
            {
                _tail.Next = null;
            }
        }

        node.Prev = null;
        node.Item = default!;
        _count--;
        return value;
    }

    // --------------------------------------------------------------------
    // Search / removal by value
    // --------------------------------------------------------------------

    /// <summary>
    /// Returns true if the list contains at least one element equal to the given value.
    /// Uses EqualityComparer&lt;T&gt;.Default for comparison.
    /// </summary>
    public bool Contains(T value)
    {
        return FindNode(value) is not null;
    }

    /// <summary>
    /// Removes the first occurrence of the given value from the list, if present.
    /// Uses EqualityComparer&lt;T&gt;.Default for comparison.
    /// </summary>
    /// <returns>true if an element was removed; false otherwise.</returns>
    public bool RemoveFirstOccurrence(T value)
    {
        var node = FindNode(value);
        if (node is null)
        {
            return false;
        }

        Unlink(node);
        return true;
    }

    private Node? FindNode(T value)
    {
        var comparer = EqualityComparer<T>.Default;
        var current = _head;

        while (current is not null)
        {
            if (comparer.Equals(current.Item, value))
            {
                return current;
            }
            current = current.Next;
        }

        return null;
    }

    /// <summary>
    /// Unlinks the given node from the list, adjusting head/tail and count.
    /// Assumes the node belongs to this list.
    /// </summary>
    private void Unlink(Node node)
    {
        var prev = node.Prev;
        var next = node.Next;

        if (prev is null)
        {
            // node is head
            _head = next;
        }
        else
        {
            prev.Next = next;
        }

        if (next is null)
        {
            // node is tail
            _tail = prev;
        }
        else
        {
            next.Prev = prev;
        }

        node.Prev = null;
        node.Next = null;
        node.Item = default!;
        _count--;
    }

    // --------------------------------------------------------------------
    // Iteration
    // --------------------------------------------------------------------

    /// <summary>
    /// Returns an enumerator that iterates through the list from first to last.
    /// </summary>
    public IEnumerator<T> GetEnumerator()
    {
        return new ForwardEnumerator(this);
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    private sealed class ForwardEnumerator : IEnumerator<T>
    {
        private readonly DoublyLinkedList<T> _list;
        private Node? _current;
        private Node? _lastReturned;

        public ForwardEnumerator(DoublyLinkedList<T> list)
        {
            _list = list;
            _current = list._head;
        }

        public T Current
        {
            get
            {
                if (_lastReturned is null)
                    throw new InvalidOperationException("Enumeration has not started. Call MoveNext.");
                return _lastReturned.Item;
            }
        }

        object IEnumerator.Current => Current!;

        public bool MoveNext()
        {
            if (_current is null)
            {
                _lastReturned = null;
                return false;
            }

            _lastReturned = _current;
            _current = _current.Next;
            return true;
        }

        public void Reset()
        {
            _current = _list._head;
            _lastReturned = null;
        }

        public void Dispose()
        {
            // nothing to dispose
        }
    }

    /// <summary>
    /// Returns an enumerable that iterates the elements in reverse order
    /// (from last to first).
    /// </summary>
    public IEnumerable<T> AsReversed()
    {
        return AsReversedIterator();

        IEnumerable<T> AsReversedIterator()
        {
            var current = _tail;
            while (current is not null)
            {
                yield return current.Item;
                current = current.Prev;
            }
        }
    }

    // --------------------------------------------------------------------
    // ToString / debug
    // --------------------------------------------------------------------

    public override string ToString()
    {
        var sb = new StringBuilder();
        sb.Append("DoublyLinkedList[");

        var current = _head;
        while (current is not null)
        {
            sb.Append(current.Item);
            if (current.Next is not null)
            {
                sb.Append(", ");
            }
            current = current.Next;
        }

        sb.Append(']');
        return sb.ToString();
    }
}
